File: rust/src/task.rs
use serde::{Deserialize, Serialize};
use std::time::Instant;

#[derive(Clone, Deserialize, Serialize, Debug)]
pub struct Task {
    pub done: bool,
    pub title: String,
    #[serde(with = "serde_millis")]
    pub created_at: Instant,
    #[serde(with = "serde_millis")]
    pub done_at: Instant,
}

impl Task {
    pub fn new(title: String) -> Result<Self, String> {
        if title.is_empty() {
            return Err("title must not be empty".to_string());
        }

        Ok(Self {
            done: false,
            title,
            created_at: Instant::now(),
            done_at: Instant::now(),
        })
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        let task = super::Task::new("test".to_string());

        if let Ok(t) = task {
            assert_eq!("test".to_string(), t.title);
        }
    }

    #[test]
    fn no_title() {
        let task = super::Task::new("".to_string());

        if let Err(e) = task {
            assert_eq!("title must not be empty".to_string(), e);
        }
    }
}

File: zig/src/task.zig
const std = @import("std");

pub const TaskCreateError = error{
    RequiredTitle,
};

pub const Task = struct {
    done: bool,
    title: []const u8,
    created_at: i64,
    done_at: i64,

    pub fn create(title: []const u8) TaskCreateError!Task {
        if (title.len == 0) {
            return error.RequiredTitle;
        }

        return Task{
            .done = false,
            .title = title,
            .created_at = std.time.timestamp(),
            .done_at = std.time.timestamp(),
        };
    }
};

test "check for task title" {
    try std.testing.expectError(error.RequiredTitle, Task.create(""));
}

test "success" {
    const t = try Task.create("A title");
    try std.testing.expectEqualSlices(u8, "A title", t.title);
}
